### ISSUE TITLE
Extend type constraints in opschema to support type promotion

### ISSUE THREAD

**Issue by titaiwangms** [microsoft/onnxscript#746]
The type constraints in opschema only limits inputs in certain range. It loses the information of type promotion. For example, in [BatchNormalization](https://github.com/onnx/onnx/blob/main/docs/Operators.md#BatchNormalization), T, T1, and T2 all have the same type constraints range, but the different name of type constraints actually indicates they should be different. Type promotion relies on this information.

Related https://github.com/microsoft/onnxscript/issues/563

cc @justinchuby @abock @BowenBao @gramalingam @thiagocrepaldi

**Comment by justinchuby** [comment]
I think the info is accessible in OpSchema? Was I missing something?

**Comment by titaiwangms** [comment]
This is referring something like #768. Do we already have that?

**Comment by justinchuby** [comment]
We only generate OpSchemas for functions (user defined ops). ONNX built in Ops don’t rely on the annotations for OpSchemas. Theirs are retrieved directly from ONNX IIRC.

**Comment by titaiwangms** [comment]
I guess I don't know how to rephrase this better. The issue is that we need T/T1 information on torchlib to tell converter that we need type promotion or not.

**Comment by justinchuby** [comment]
Was the information in the FormalParameter s in the OpSchema? You should be able to see the name and the type_str of the param. (I should publish documentation for the class in ONNX).

**Comment by titaiwangms** [comment]
I did find T/T1/T2 in OpSchema, but I thought it was not meaningful, as we manually annotated type. And I don't recall we consider T info at the time?

**Comment by titaiwangms** [comment]
Based on offline chat, this could totally be resolved if we utilize typing system strictly in torchlib:

We take aten::native_batch_norm for example,

What we have right now is:
```python
@torch_op("aten::native_batch_norm", trace_only=True)
def aten_native_batch_norm(
    input: TFloat,
    weight: Optional[TFloat] = None,
    bias: Optional[TFloat] = None,
    running_mean: Optional[TFloat] = None,
    running_var: Optional[TFloat] = None,
    training: bool = False,
    momentum: float = 0.9,
    eps: float = 1e-05,
) -> Tuple[TFloat, TFloat, TFloat]:
    ...
```

However, we should type it as:
```python
@torch_op("aten::native_batch_norm", trace_only=True)
def aten_native_batch_norm(
    input: TFloat,
    weight: Optional[TFloat2] = None,
    bias: Optional[TFloat2] = None,
    running_mean: Optional[TFloat3] = None,
    running_var: Optional[TFloat3] = None,
    training: bool = False,
    momentum: float = 0.9,
    eps: float = 1e-05,
) -> Tuple[TFloat, TFloat, TFloat]:
    ...
```

By using TFloatN, we can annotate and seperate the T info in the current codebase
